File: Sources/MoonRuntime/WorkerPool.swift
import Foundation

public enum ModelTier: String, Sendable {
    case flash
    case pro
}

public struct WorkerPoolConfig: Sendable {
    public var flashConcurrency: Int
    public var proConcurrency: Int
    public var onAcquire: (@Sendable (ModelTier, Int) -> Void)?

    public init(
        flashConcurrency: Int = 20,
        proConcurrency: Int = 5,
        onAcquire: (@Sendable (ModelTier, Int) -> Void)? = nil
    ) {
        self.flashConcurrency = flashConcurrency
        self.proConcurrency = proConcurrency
        self.onAcquire = onAcquire
    }
}

public final class WorkerPool: @unchecked Sendable {
    private let limits: WorkerPoolConfig
    private var active: [ModelTier: Int] = [.flash: 0, .pro: 0]
    private var queues: [ModelTier: [CheckedContinuation<Void, Never>]] = [.flash: [], .pro: []]
    private let lock = NSLock()

    public init(config: WorkerPoolConfig = WorkerPoolConfig()) {
        self.limits = config
    }

    public func run<T: Sendable>(_ tier: ModelTier, _ fn: @Sendable () async throws -> T) async rethrows -> T {
        await acquire(tier)
        defer { release(tier) }
        return try await fn()
    }

    public func runAll<T: Sendable>(_ tier: ModelTier, _ tasks: [@Sendable () async throws -> T]) async throws -> [T] {
        try await withThrowingTaskGroup(of: (Int, T).self) { group in
            for (index, task) in tasks.enumerated() {
                group.addTask {
                    let value = try await self.run(tier, task)
                    return (index, value)
                }
            }

            var results = Array<T?>(repeating: nil, count: tasks.count)
            for try await (index, value) in group {
                results[index] = value
            }
            return results.map { $0! }
        }
    }

    private func acquire(_ tier: ModelTier) async {
        let shouldWait: Bool = lock.withLock {
            let limit = tier == .flash ? limits.flashConcurrency : limits.proConcurrency
            let current = active[tier, default: 0]
            if current < limit {
                bumpLocked(tier)
                return false
            }
            return true
        }

        if shouldWait {
            await withCheckedContinuation { continuation in
                lock.withLock {
                    queues[tier, default: []].append(continuation)
                }
            }
        }
    }

    private func bumpLocked(_ tier: ModelTier) {
        let next = active[tier, default: 0] + 1
        active[tier] = next
        limits.onAcquire?(tier, next)
    }

    private func release(_ tier: ModelTier) {
        let waiter: CheckedContinuation<Void, Never>? = lock.withLock {
            let next = max(0, active[tier, default: 0] - 1)
            active[tier] = next
            if var queue = queues[tier], !queue.isEmpty {
                let nextWaiter = queue.removeFirst()
                queues[tier] = queue
                bumpLocked(tier)
                return nextWaiter
            }
            return nil
        }
        waiter?.resume()
    }
}

public func modelToTier(_ model: String) -> ModelTier {
    model.contains("flash") ? .flash : .pro
}

private extension NSLock {
    func withLock<T>(_ body: () throws -> T) rethrows -> T {
        lock()
        defer { unlock() }
        return try body()
    }
}

File: legacy/packages/runtime/src/worker-pool.ts
export type ModelTier = "flash" | "pro";

export interface WorkerPoolConfig {
  flashConcurrency: number;
  proConcurrency: number;
  onAcquire?: (tier: ModelTier, active: number) => void;
}

const DEFAULT_CONFIG: WorkerPoolConfig = {
  flashConcurrency: 20,
  proConcurrency: 5,
};

export class WorkerPool {
  private readonly limits: WorkerPoolConfig;
  private readonly active = new Map<ModelTier, number>();
  private readonly queues = new Map<ModelTier, Array<() => void>>();

  constructor(config: Partial<WorkerPoolConfig> = {}) {
    this.limits = { ...DEFAULT_CONFIG, ...config };
    this.active.set("flash", 0);
    this.active.set("pro", 0);
    this.queues.set("flash", []);
    this.queues.set("pro", []);
  }

  async run<T>(tier: ModelTier, fn: () => Promise<T>): Promise<T> {
    await this.acquire(tier);
    try {
      return await fn();
    } finally {
      this.release(tier);
    }
  }

  async runAll<T>(tier: ModelTier, tasks: (() => Promise<T>)[]): Promise<T[]> {
    return Promise.all(tasks.map((task) => this.run(tier, task)));
  }

  private async acquire(tier: ModelTier): Promise<void> {
    const limit = tier === "flash" ? this.limits.flashConcurrency : this.limits.proConcurrency;

    if ((this.active.get(tier) ?? 0) < limit) {
      this.bump(tier);
      return;
    }

    await new Promise<void>((resolve) => {
      this.queues.get(tier)!.push(() => {
        this.bump(tier);
        resolve();
      });
    });
  }

  private bump(tier: ModelTier): void {
    const next = (this.active.get(tier) ?? 0) + 1;
    this.active.set(tier, next);
    this.limits.onAcquire?.(tier, next);
  }

  private release(tier: ModelTier): void {
    const next = Math.max(0, (this.active.get(tier) ?? 0) - 1);
    this.active.set(tier, next);

    const queue = this.queues.get(tier)!;
    const waiter = queue.shift();
    if (waiter) waiter();
  }
}

export function modelToTier(model: string): ModelTier {
  return model.includes("flash") ? "flash" : "pro";
}